jina/peapods/peas/__init__.py

```diff
@@ -1,338 +1,174 @@
+            raise RuntimeFailToStart from ex
 
-    @property
-    def request(self) -> 'Request':
-        """Get the current request body inside the protobuf message"""
-        return self._request
+    def run(self):
+        """ Method representing the :class:`BaseRuntime` activity.
 
-    @property
-    def message(self) -> 'Message':
-        """Get the current protobuf message to be processed"""
-        return self._message
+        .. note::
+            :meth:`run` is running in subprocess/thread, the exception can not be propagated to the main process.
+            Hence, please do not raise any exception here.
+        """
 
-    @property
-    def request_type(self) -> str:
-        """Get the type of message being processed"""
-        return self._message.envelope.request_type
+        def _finally():
+            self.is_shutdown.set()
+            self.is_ready.clear()
+            self._unset_envs()
 
-    def load_executor(self):
-        """Load the executor to this BasePea, specified by ``uses`` CLI argument.
+        self._set_envs()
 
-        """
         try:
-            try:
-                self.executor = BaseExecutor.load_config(self.args.uses,
-                                                         separated_workspace=self.args.separated_workspace,
-                                                         pea_id=self.args.pea_id,
-                                                         read_only=self.args.read_only)
-            except BadConfigSource:
-                # retry loading but with "uses_internal" as the source
-                self.executor = BaseExecutor.load_config(self.args.uses_internal,
-                                                         separated_workspace=self.args.separated_workspace,
-                                                         pea_id=self.args.pea_id,
-                                                         read_only=self.args.read_only)
-            self.executor.attach(pea=self)
-        except FileNotFoundError as ex:
-            self.logger.error(f'fail to load file dependency: {repr(ex)}')
-            raise ExecutorFailToLoad from ex
+            self.runtime.setup()
         except Exception as ex:
-            raise ExecutorFailToLoad from ex
-
-    def print_stats(self):
-        self.logger.info(
-            ' '.join(f'{k}: {v / self._timer.accum_time["loop"]:.2f}' for k, v in self._timer.accum_time.items()))
-
-    def save_executor(self):
-        """Save the contained executor according to the `dump_interval` parameter
-        """
-        if (time.perf_counter() - self.last_dump_time) > self.args.dump_interval > 0:
-            self.executor.save()
-            self.last_dump_time = time.perf_counter()
-            if hasattr(self, 'zmqlet'):
-                self.zmqlet.print_stats()
+            self.logger.error(f'{ex!r} during {self.runtime.setup!r}', exc_info=self.args.show_exc_info)
+        else:
+            self.is_ready.set()
```

Comment:
this else construction isnt it a bit weird?

Reply:
The use of the else clause is better than adding additional code to the try clause because it avoids accidentally catching an exception that wasn’t raised by the code being protected by the try ... except statement.

Reply:
good point

Reply:
one of the big benefits of the four-function-design of `BaseRuntime` is its decoupling the progress-status of the Runtime: for example, you never have to inject `is_ready` from `Pea` into some `Runtime.func()` just to let `Pea` to know `Runtime` is ready. 

Therefore, try except at `BasePea` must catch exception in a precise way. It must see the four functions from `Runtime` as the atomic function and catch each of them precisely.